Make structural copy handle Blists properly.

markuspf opened the pull request:
This is a rebased and updated version of #1455.

Sorry I couldn't keep my fingers off the `clang-format` button.

fingolfin commented:
This looks good now. Hmm... perhaps add a test case, based on #1428 ? (I'd just add a test file for `StructuralCopy` and copy the test cases from the issue into it, adjusted. We can add more `StructuralCopy` tests in the future.

markuspf commented:
Tests added.

fingolfin commented on src/blister.c:
```diff
@@ -284,17 +284,30 @@ Obj DoCopyBlist(Obj list, Int mut) {
-    /* don't change immutable objects                                      */
-    if ( ! IS_MUTABLE_OBJ(list) ) {
-        return list;
-    }
```
Huh... Hadn't thought about this before, but shouldn't this check for (im)mutable stay in? It makes sense to me from a logical point of view, and I just checked, this is also what we do for plists. @stevelinton any reason why your original PR removed this?!?

markuspf replied:
I added the code in again as to make this sensibly mergeable, because I agree that the check should be there. If @stevelinton comments and gives a reason why it can go, I can remove this again.

stevelinton replied:
Because immutable Blists now have their own function entered appropriately in the dispatch table. Perhaps `assert(IS_MUTABLE_OBJ(list));` should be added here.

fingolfin commented on tst/testinstall/opers/StructuralCopy.tst:
```diff
@@ -0,0 +1,23 @@
+# Blist
+gap> a:=[true];; IsBlistRep(a);
```
Add START_TEST / STOP_TEST to this .tst file, like in the other files in that dir?

markuspf commented:
I removed the check for `!IS_MUTABLE_OBJ` again and added a `GAP_ASSERT(IS_MUTABLE_OBJ(list))`. While I was there I also added a `GAP_ASSERT(!IS_MUTABLE_OBJ(list))` into the handler for immutable blists.

Therefore, when tests pass we should be good to go.
